clients/src/main/java/org/apache/kafka/clients/admin/ElectPreferredLeadersResult.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients.admin;
+
+import org.apache.kafka.common.KafkaFuture;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.annotation.InterfaceStability;
+import org.apache.kafka.common.internals.KafkaFutureImpl;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * The result of {@link AdminClient#electPreferredLeaders(Collection, ElectPreferredLeadersOptions)}
+ *
+ * The API of this class is evolving, see {@link AdminClient} for details.
+ */
+@InterfaceStability.Evolving
+public class ElectPreferredLeadersResult {
+
+    private final KafkaFuture<? extends Map<TopicPartition, ? extends KafkaFuture<Void>>> futures;
+
+    ElectPreferredLeadersResult(KafkaFuture<? extends Map<TopicPartition, ? extends KafkaFuture<Void>>> futures) {
+        this.futures = futures;
+    }
+
+    /** Return a new future that has completed exceptionally */
+    private static <T> KafkaFuture<T> exceptionalFuture(Throwable e) {
+        KafkaFutureImpl<T> future = new KafkaFutureImpl<>();
+        future.completeExceptionally(e);
+        return future;
+    }
+
+    /**
+     * Get the result of the election for the given {@code partition}.
+     * If there was not an election triggered for the given {@code partition}, the
+     * returned future will complete with an error.
+     */
+    public KafkaFuture<Void> partitionResult(TopicPartition partition) {
+        final Map<TopicPartition, ? extends KafkaFuture<Void>> map;
+        try {
+            map = futures.get();
```

Comment:
You are calling KafkaFuture#get here, which is blocking.  Don't do that.

Reply:
@cmccabe You're right, but to fix I'll need to implement a `KafkaFuture.thenCompose(Function<T, KafkaFuture<R>>)` method on `KafkaFuture<T>`. That seems to be awkward because the API of KafkaFuture doesn't provide a public `addWaiter()` method, but the `Function<A, KafkaFuture<B>>` in the API of `thenApply()` is returning a `KafkaFuture`, not an Impl. I've pushed  WIP patch, so you can see what I mean. Thoughts?

Reply:
I don't see why you need `thenCompose`.  You already have `thenApply`, and you can just do something like this:
```
final KafkaFuture<Map<TopicPartition, Throwable>> partitionFuture = new KafkaFutureImpl<>();
rpcFuture.thenApply(new Function<Map<TopicPartition, Throwable>, Void>() {
  @Override
  Void apply(Map<TopicPartition, Throwable> map) {
    // check map for this partition and ensure there is no error
   if (not found in map) {
     partitionFuture.completeExceptionally(not there);
   else if (there was some error) {
     partitionFuture.completeExceptionally(error);
   } else {
     partitionFuture.complete(null);
   }
  }
});
```

I am not opposed to adding `thenCompose` to KafkaFuture, but I just don't think you need it for this.

Reply:
@cmccabe that code makes no sense to me. Is it supposed to be a replacement for what's in `ElectPreferredLeadersResult.partitionResult()`? What's supposed to happen with your `partitionFuture`? It gets completed, but doesn't otherwise get used/returned/passed to anything, so nothing can handle or wait on its result. Your `apply()` lacks a `return`, if it always returns null then the result of the `thenApply()` is a `Future<Void>` that is always successful.

I think we need `thenCompose()` because, in effect, we want a `Function` that can return a `Future<Void>`. `thenApply()` forces us into returning the actual `Void`, but then we are having to call `get()` in the `Function`, which defeats the point. Or we can use `thenApply()` and return a `Future<Void>` from the `Function`, but then the result of `thenApply()` is a double `Future<Future<Void>>`.

Reply:
@cmccabe I need your help to resolve this.